### PR TITLE
Scroll cursor and page together (neovim-like scrolling)

### LINKED ISSUE #7639: Neovim-like exact scrolling
I find helix's keyboard scrolling (using C-d/C-u for example) quite disorientating mainly because I often want to scroll up a bit, then scroll back down to the cursor's original position, helix's scrolling doesn't always do this depending on initial cursor position. This is the behaviour I want (exact same behaviour as neovim).

https://github.com/helix-editor/helix/assets/75994927/18f4562f-c9ea-4de0-8a16-ee4cc8ebcdf4

I dont really like this behaviour for mouse scrolling though. 

Would you support this as an option in the config? something like `keyboard-scroll-exact: true | false`, `mouse-scroll-exact: true | false`. Or maybe only have the keyboard option

I've implemented this behaviour locally and can make a pr for this

### PR DESCRIPTION
**Pull request by AlexanderDickie**
Solves #7639 - adds a new scroll function so that the cursor and page jump by the same distance vertically. Assigns this to C-u and C-d, similar to neovim. 

The consistent distance between cursor and view top would break if the cursor were to land on a whole line of virtual text- the cursor would be placed on the closest non-virtual text above where it landed. Meaning the view and cursor would not have jumped the same vertical distance. 
This should be pretty rare though- unless you are using #6417 and have a ton of error messages. I could refactor this nvim-scroll function though- to first calculate how many lines the cursor jumped (eg the cursor tried to jump by half a page but could only jump by half a page -1 line due to virtual text) and then move the view by exactly this number. This would change the behaviour at the bottom of the page a bit also.

This nvim-scroll function respects the cursor's column throughout scrolls which I think is nice, I will make a pr for the `scroll` function to have the same.

Closes #7639

### CODE REVIEW

File: helix-core/src/movement.rs
Comment (AlexanderDickie) [reply helix-core/src/movement.rs:67]: I want the cursor and page to jump by the same amount of vertical distance on each scroll. 
My function calls this method to move the cursor, if this line wasnt removed it would be moving the cursor a distance of `count` without respect for complete lines of virtual text (inline diagnostics) (eg which are present in #6417 ), meaning it would be jumping a greater visual distance that the view is. (As the view calls `char_idx_at_visual_offset` which respects whole lines of virtual text).

I've instead called this line in `move_impl` so that functions like `move_visual_line_down` will not land the cursor on complete lines of visual text/inlnie diagnostics.

Comment (pascalkuthe) [reply helix-core/src/movement.rs:67]: You probably need to handle that differently. There is a fast oath for the case that softwrap is disabled and it doesn't respect multilinear virtual text anyway (it just uses line numbers).

Comment (AlexanderDickie) [reply helix-core/src/movement.rs:67]: a fast oath?

Comment (pascalkuthe) [reply helix-core/src/movement.rs:67]: https://github.com/helix-editor/helix/blob/01a1e5ec2a8c5b81da1360746dad0a96cd88ae8e/helix-core/src/movement.rs#L64

Comment (AlexanderDickie) [reply helix-core/src/movement.rs:67]: like this [e490572](https://github.com/helix-editor/helix/pull/8015/commits/e4905729c338a2260e6981f1d8fac022897b4191)? I am getting inconsistent jumping when the cursor starts on the non-virtual line directly above a multiline diagnostic, the scroll up is inconsistent depending on which word in the line the cursor starts on. Not sure what is causing this

Comment (pascalkuthe) [reply helix-core/src/movement.rs:67]: yeah, that should work. Honestly, I think for now you can just use the existing `move_vertically_visual` function without modifying it. I will look into this when I get back to the inline diagnostic PR. That PR still has many bugs so its not clear if that is actually a problem with you code.

Comment (AlexanderDickie) [reply helix-core/src/movement.rs]: ok [3cc3e2a](https://github.com/helix-editor/helix/pull/8015/commits/3cc3e2a489a7f880fa3ffae7fade441289f95778), this small refactor does also allow me to easily solve the 'constant distance between view and cursor if cursor lands on whole line of virtual text' that i was talking about in the description. But this can also wait until the inline diagnostics pr gets fixed/merged too.

File: helix-term/src/commands.rs
Comment (pascalkuthe) [inline helix-term/src/commands.rs:1510]: this has quite a lot of overhead (its never cached here) I don't think we should we special case the scrolloff this way. vim doesn't do that either

File: helix-term/src/commands.rs
Comment (pascalkuthe) [inline helix-term/src/commands.rs:1540]: can we share the code between the normal scrolling and this scroll function to avoid code duplication. I have some further changes planned here and I don't want to do them twice

File: helix-term/src/commands.rs
Comment (AlexanderDickie) [reply helix-term/src/commands.rs:1550]: [949cfb9](https://github.com/helix-editor/helix/pull/8015/commits/949cfb98fb72c29ee3f401d41ae3ae6d09c67669)

File: helix-term/src/commands.rs
Comment (AlexanderDickie) [reply helix-term/src/commands.rs:1510]: As in disregard the scrolloff- just like `move_visual_line_down` does?
This will mean that when we do a C-d from the line 1, we wont do an exact half page scroll in terms of the view. But neovim will actually do an exact half page scrolls in terms of the view

Comment (pascalkuthe) [reply helix-term/src/commands.rs:1510]: hmm I see what you mean, then you need to write your own version of `move_vertically_visual`. That function already calculates the vertical position of the cursor so doing it twice is very inefficient

File: helix-term/src/commands.rs
Comment (AlexanderDickie) [reply helix-term/src/commands.rs]: [5a48de0](https://github.com/helix-editor/helix/pull/8015/commits/5a48de0800b4433cf1b8d9bb41bf74d1b5d3a11c)

File: helix-term/src/commands.rs
Comment (AlexanderDickie) [reply helix-term/src/commands.rs]: yeh I agree making the scroll distance consistent is probably better- [5a48de0](https://github.com/helix-editor/helix/pull/8015/commits/5a48de0800b4433cf1b8d9bb41bf74d1b5d3a11c)

**Review by pascalkuthe — DISMISSED** [review]
very nice this is now a very small/clean change that  adds something that I find very useful. Thanks for working on this!

**Comment by pascalkuthe** [comment]
looks like formatting still needs to be fixed tough

**Comment by pascalkuthe** [comment]
(and some clippy lints)

**Review by pascalkuthe — APPROVED** [review]

**Comment by wildwestrom** [comment]
Does this require the documentation to be updated?

**Comment by AlexanderDickie** [comment]
Can we get the labels changed on this so it gets reviewed? 😄

**Comment by pascalkuthe** [comment]
The labels don't really affect what we review too much. It's mostly for other people

**Review by the-mikedavis — APPROVED** [review]

**Review by pascalkuthe — APPROVED** [review]

**Pull request merged**
